File: chapters/08_queue_and_concurrency_lanes/v3_lane_concurrency.py
from __future__ import annotations

import asyncio


async def _run_one(task_id: str, delay: float = 0.01) -> str:
    await asyncio.sleep(delay)
    return f"done:{task_id}"


def _parse_task(task: tuple[str, str] | tuple[str, str, float]) -> tuple[str, str, float]:
    if len(task) == 2:
        task_id, lane = task
        return task_id, lane, 0.01
    task_id, lane, delay = task
    return task_id, lane, delay


async def run_with_lanes(
    tasks: list[tuple[str, str] | tuple[str, str, float]],
    lane_limit: int = 1,
    global_limit: int = 2,
) -> list[str]:
    results, _ = await run_with_lanes_trace(tasks, lane_limit=lane_limit, global_limit=global_limit)
    return results


async def run_with_lanes_trace(
    tasks: list[tuple[str, str] | tuple[str, str, float]],
    lane_limit: int = 1,
    global_limit: int = 2,
) -> tuple[list[str], list[str]]:
    lane_locks: dict[str, asyncio.Semaphore] = {}
    global_lock = asyncio.Semaphore(global_limit)
    trace: list[str] = []

    async def run_task(task_id: str, lane: str, delay: float) -> str:
        trace.append(f"queued:{task_id}:{lane}")
        lane_lock = lane_locks.setdefault(lane, asyncio.Semaphore(lane_limit))
        async with global_lock:
            trace.append(f"enter-global:{task_id}:{lane}")
            async with lane_lock:
                trace.append(f"enter-lane:{task_id}:{lane}")
                out = await _run_one(task_id, delay=delay)
                trace.append(f"exit-lane:{task_id}:{lane}")
                return out

    coros = []
    for item in tasks:
        task_id, lane, delay = _parse_task(item)
        coros.append(run_task(task_id, lane, delay))
    return list(await asyncio.gather(*coros)), trace


def main() -> None:
    tasks = [("A", "s1"), ("B", "s1"), ("C", "s2")]
    print(asyncio.run(run_with_lanes(tasks, lane_limit=1, global_limit=2)))


if __name__ == "__main__":
    main()

File: chapters/08_queue_and_concurrency_lanes/v3_lane_concurrency.ts
// TypeScript mirror for Chapter 08: Queue and Concurrency Lanes (v3)
// Run with: npx ts-node v3_lane_concurrency.ts

// Note: This uses async/await patterns compatible with Node.js

type Task2 = [string, string];
type Task3 = [string, string, number];
type Task = Task2 | Task3;

interface Semaphore {
  permits: number;
  queue: (() => void)[];
}

function createSemaphore(permits: number): Semaphore {
  return { permits, queue: [] };
}

async function acquire(sem: Semaphore): Promise<void> {
  if (sem.permits > 0) {
    sem.permits--;
    return;
  }
  return new Promise((resolve) => {
    sem.queue.push(resolve);
  });
}

function release(sem: Semaphore): void {
  if (sem.queue.length > 0) {
    const next = sem.queue.shift()!;
    next();
  } else {
    sem.permits++;
  }
}

async function runOne(taskId: string, delay: number): Promise<string> {
  await new Promise((r) => setTimeout(r, delay * 1000));
  return `done:${taskId}`;
}

function parseTask(task: Task): [string, string, number] {
  if (task.length === 2) {
    const [taskId, lane] = task;
    return [taskId, lane, 0.01];
  }
  return task;
}

async function runWithLanesTrace(
  tasks: Task[],
  laneLimit: number = 1,
  globalLimit: number = 2
): Promise<[string[], string[]]> {
  const laneLocks: Map<string, Semaphore> = new Map();
  const globalLock = createSemaphore(globalLimit);
  const trace: string[] = [];

  async function runTask(taskId: string, lane: string, delay: number): Promise<string> {
    trace.push(`queued:${taskId}:${lane}`);
    if (!laneLocks.has(lane)) {
      laneLocks.set(lane, createSemaphore(laneLimit));
    }
    const laneLock = laneLocks.get(lane)!;

    await acquire(globalLock);
    trace.push(`enter-global:${taskId}:${lane}`);

    await acquire(laneLock);
    trace.push(`enter-lane:${taskId}:${lane}`);

    const out = await runOne(taskId, delay);

    trace.push(`exit-lane:${taskId}:${lane}`);
    release(laneLock);
    release(globalLock);

    return out;
  }

  const coros = tasks.map((t) => {
    const [taskId, lane, delay] = parseTask(t);
    return runTask(taskId, lane, delay);
  });

  const results = await Promise.all(coros);
  return [results, trace];
}

async function main(): Promise<void> {
  const tasks: Task[] = [["A", "s1"], ["B", "s1"], ["C", "s2"]];
  const [results, trace] = await runWithLanesTrace(tasks, 1, 2);
  console.log(results);
  console.log(trace);
}

main();
